libs/dyn/pattern_trie.go

```diff
@@ -0,0 +1,182 @@
+	if len(pattern) == 0 {
+		t.root.isEnd = true
+		return nil
+	}
+
+	current := t.root
+	for i, component := range pattern {
+		// Create next node based on component type
+		var next *trieNode
+		switch c := component.(type) {
+		case anyKeyComponent:
+			if current.anyKey == nil {
+				current.anyKey = &trieNode{}
+			}
+			next = current.anyKey
+
+		case anyIndexComponent:
+			if current.anyIndex == nil {
+				current.anyIndex = &trieNode{}
+			}
+			next = current.anyIndex
+
+		case pathComponent:
+			if key := c.Key(); key != "" {
+				if current.pathKey == nil {
+					current.pathKey = make(map[string]*trieNode)
+				}
+				if _, exists := current.pathKey[key]; !exists {
+					current.pathKey[key] = &trieNode{}
+				}
+				next = current.pathKey[key]
+			} else {
+				idx := c.Index()
+				if current.pathIndex == nil {
+					current.pathIndex = make(map[int]*trieNode)
+				}
+				if _, exists := current.pathIndex[idx]; !exists {
+					current.pathIndex[idx] = &trieNode{}
+				}
+				next = current.pathIndex[idx]
+			}
+		}
+
+		if next == nil {
+			return fmt.Errorf("invalid component type: %T", component)
+		}
+
+		// Mark as end of pattern if this is the last component.
+		if !next.isEnd && i == len(pattern)-1 {
+			next.isEnd = true
+		}
+
+		// Move to next node
+		current = next
+	}
+
+	return nil
+}
+
+// SearchPath checks if the given path matches any pattern in the trie.
+// A path matches if it exactly matches a pattern or if it matches a pattern with wildcards.
+func (t *PatternTrie) SearchPath(path Path) (Pattern, bool) {
+	// We statically allocate the prefix array that is used to track the current
+	// prefix accumulated while walking the prefix tree. Having the static allocation
+	// ensures that we do not allocate memory on every recursive call.
+	prefix := make(Pattern, len(path))
+	pattern, ok := t.searchPathRecursive(t.root, path, prefix, 0)
+	return pattern, ok
+}
+
+// searchPathRecursive is a helper function that recursively checks if a path matches any pattern.
+// Arguments:
+// - node: the current node in the trie.
+// - path: the path to check.
+// - prefix: the prefix accumulated while walking the prefix tree.
+// - index: the current index in the path / prefix
+//
+// Note we always expect the path and prefix to be the same length because wildcards like * and [*]
+// only match a single path component.
+func (t *PatternTrie) searchPathRecursive(node *trieNode, path Path, prefix Pattern, index int) (Pattern, bool) {
+	if node == nil {
+		return nil, false
+	}
+
+	// Zero case, when the query path is the root node. We return nil here to match
+	// the semantics of [MustPatternFromString] which returns nil for the empty string.
+	if len(path) == 0 {
+		return nil, node.isEnd
+	}
+
+	// If we've reached the end of the path, check if this node is a valid end of a pattern.
+	isLast := index == len(path)
+	if isLast {
+		return prefix, node.isEnd
+	}
+
+	currentComponent := path[index]
```

Comment:
Do we have a requirement to handle Unicode in paths? This code will most likely fail with multi-byte characters in paths